rust/kernel/static_assert.rs

```diff
@@ -4,35 +4,198 @@
+/// allowed to easily manipulate integers: no 128-bit code will be generated since it will be
+/// evaluated in a const context.
+///
+/// ## Examples
+///
+/// ```
+/// // Trivial usage.
+/// static_assert!(( -1) does not fit  in u8);
+/// static_assert!((  0)          fits in u8);
+/// static_assert!((255)          fits in u8);
+/// static_assert!((256) does not fit  in u8);
+///
+/// // Two's complement.
+/// static_assert!((-128)          fits in i8);
+/// static_assert!(( 127)          fits in i8);
+/// static_assert!(( 128) does not fit  in i8);
+///
+/// // Using `i128` for easy manipulation of integers.
+/// const MAX_ERRNO: u32 = 4095;
+/// static_assert!((-(MAX_ERRNO as i128)) fits in i16);
+/// ```
 #[macro_export]
 macro_rules! static_assert {
+    // Boolean assertion: `expr`.
     ($condition:expr) => {
         // Based on the latest one in `rustc`'s one before it was [removed].
         //
         // [removed]: https://github.com/rust-lang/rust/commit/c2dad1c6b9f9636198d7c561b47a2974f5103f6d
         #[allow(dead_code)]
         const _: () = [()][!($condition) as usize];
     };
+
+    // Set membership assertion: `(expr) is in {a0, ..., aN}`.
+    (($expression:expr) is in {$($a:expr),+}) => {
+        static_assert!(  $(($expression) == ($a))||* );
+    };
+    (($expression:expr) is not in {$($a:expr),+}) => {
+        static_assert!(!($(($expression) == ($a))||*));
+    };
+
+    // Interval membership assertion: `(expr) is in [min, max]`.
+    (($expression:expr) is in [$min:expr, $max:expr]) => {
+        static_assert!(  ($expression) >= ($min) && ($expression) <= ($max) );
+    };
+    (($expression:expr) is not in [$min:expr, $max:expr]) => {
+        static_assert!(!(($expression) >= ($min) && ($expression) <= ($max)));
+    };
+
+    // Fits-in-type assertion: `(expr) fits in type`.
+    (($expression:expr) fits in $t:ty) => {
+        static_assert!(($expression) is     in [<$t>::MIN as i128, <$t>::MAX as i128]);
+    };
+    (($expression:expr) does not fit in $t:ty) => {
+        static_assert!(($expression) is not in [<$t>::MIN as i128, <$t>::MAX as i128]);
+    };
+}
+
+// Tests.
+//
+// These should later on go into a proper test.
+
+static_assert!(42 > 24);
+static_assert!(core::mem::size_of::<u8>() == 1);
+
+const X: &[u8] = b"bar";
+static_assert!(X[1] == b'a');
+
+const fn f(x: i32) -> i32 {
+    x + 2
 }
+static_assert!(f(40) == 42);
+
+static_assert!((-2) is not in {-1, 0, 2});
+static_assert!((-1) is     in {-1, 0, 2});
+static_assert!(( 0) is     in {-1, 0, 2});
+static_assert!(( 1) is not in {-1, 0, 2});
+static_assert!(( 2) is     in {-1, 0, 2});
+static_assert!(( 3) is not in {-1, 0, 2});
+
+static_assert!((core::mem::size_of::<usize>()) is in {4, 8});
+
+static_assert!((-2) is not in [-1, 2]);
+static_assert!((-1) is     in [-1, 2]);
+static_assert!(( 0) is     in [-1, 2]);
+static_assert!(( 1) is     in [-1, 2]);
```

Comment:
Re: using ranges here:
Reading this line in isolation and without prior knowledge of this macro I would probably expect this to fail as "1 is neither -1 nor 2"